### PR TITLE
Refactor multisignature, to use index based.

### PR DESCRIPTION
**Pull request by vldm**
First iteration of refactor, to use validator id instead of public key.

### CODE REVIEW

File: blockchain/src/blockchain.rs [hunk 1]
```diff
@@ -471,6 +471,12 @@ impl Blockchain {
         self.view_change
     }
 
+    /// Returns number of total slots in current epoch.
+    /// Internally always return cfg.max_slot_count
+    pub fn total_slots(&self) -> i64 {
+        self.cfg.max_slot_count as i64
```
Comment (rtsisyk) [inline blockchain/src/blockchain.rs]: Why cast is needed here? Please either change max_slot_count to i64 or use u64 everywhere.

File: blockchain/src/error.rs [hunk 2]
```diff
@@ -55,6 +55,8 @@ pub enum BlockchainError {
     MissingValidators,
     #[fail(display = "The leader must be validator.")]
     LeaderIsNotValidator,
+    #[fail(display = "No leader signature was found in BLS signature.")]
+    NoLeaderSignatureFound,
```
Comment (rtsisyk) [inline blockchain/src/error.rs]: I suggest to change `check_multi_signature()` to return `Result<(), Error>`

File: blockchain/src/multisignature.rs [hunk 3]
```diff
@@ -43,92 +45,98 @@ pub fn create_multi_signature(
...
-        count += 1;
-    }
+    for bit in multisigmap.iter() {
+        let validator = &validators[bit];
+        let pkey: secure::G2 = validator.0.into();
+        let slots = validator.1;
+        assert!(slots > 0);
 
-    // Multi-signature must contain leader's key.
-    if !has_leader {
```
Comment (rtsisyk) [inline blockchain/src/multisignature.rs]: I think you can pass leader_id to this function and remove this part from validate_key_block():
```

            if let Some(leader_id) = validators_map.get(&leader) {
                // bit of leader should be always set.
                ensure!(
                    block.body.multisigmap.contains(*leader_id as usize),
                    BlockchainError::NoLeaderSignatureFound
                );
            } else {
                return Err(BlockchainError::LeaderIsNotValidator.into());
            }
```

File: blockchain/src/blockchain.rs [hunk 4]
```diff
@@ -563,24 +569,35 @@ impl Blockchain {
...
-            if !check_multi_signature(
-                &block_hash,
-                &block.body.multisig,
-                &block.body.multisigmap,
-                &validators,
-                &leader,
-                is_proposal,
-            ) {
+            // ignore multisignature if it proposal.
+            if !is_proposal
```
Comment (rtsisyk) [inline blockchain/src/blockchain.rs]: Please check proposal for the leader's signature.

File: blockchain/src/multisignature.rs (hunk 3 above)
Comment (vldm) [reply blockchain/src/multisignature.rs]: I think checking leader is out of responsible for multisignature.

**Pull request merged**